### PR TITLE
640 refactor model layers

### LINKED ISSUE #640: Refactor model layers
## Subject of the feature

Regarding layers and layer utils https://github.com/DeepRegNet/DeepReg/tree/main/deepreg/model, there are multiple things to optimize:

- each layer need to implement `get_config`.
- some functions in layer util should be implemented into layers

The concern here will be, whether these changes will make the checkpoint invalid again.

### PR DESCRIPTION
**Pull request by mathpluscode**
# Description

Simplify and refactor custom layer definitions.

Fixes #640

## Type of change

What types of changes does your code introduce to DeepReg?

_Please check the boxes that apply after submitting the pull request._

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Documentation Update (fix or improvement on the documentation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Please check the boxes that apply after submitting the pull request._

_If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [x] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      using `pre-commit install` and formatted all changed files. If you are not
      certain, run `pre-commit run --all-files`.
- [x] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [x] I have updated the
      [change log file](https://github.com/DeepRegNet/DeepReg/blob/main/CHANGELOG.md)
      regarding my changes.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by mathpluscode** [comment]
@YipengHu @NMontanaBrown @kate-sann5100 This PR should be ready now, please have a review if you have time ;)

The changes are major refactoring of models and layers so that
- we define fewer custom layers and all layers have `get_config` func following tf standards.
- U-Net is more general than before by providing funcs to replace different components in U-Net
- Local-Net is now a subclass of U-Net by overwriting some components.
- Global-Net is now a subclass of Local-Net without decoder/up-sampling part

File: config/test/affine.yaml [hunk 1]
```diff
@@ -3,7 +3,7 @@ train:
   backbone:
     name: "global"
     num_channel_initial: 1
-    extract_levels: [0, 1, 2, 3, 4]
+    depth: 4
```
Comment (NMontanaBrown) [inline config/test/affine.yaml:6]: does this modify back compatibility? because extract levels didn't need to be sequential before, this change suggests that all up to depth 4 is extracted?

Comment (YipengHu) [inline config/test/affine.yaml:6]: this is not clear and over-constrained - user may want only high-res layers OR only low-res layers

File: deepreg/dataset/preprocess.py [hunk 2]
```diff
@@ -326,3 +324,158 @@ def resize_inputs(
...
+        [[x' y' z' 1]] = [[x y z 1]] * [[* * * 0]
+                                        [* * * 0]
+                                        [* * * 0]
+                                        [* * * 1]]
+
+    where each * represents a degree of freedom,
+    so there are in total 12 degrees of freedom
+    the equation can be denoted as
+
+        new = old * T
```
Comment (NMontanaBrown) [inline deepreg/dataset/preprocess.py:348]: left versus right multiplication?

File: CHANGELOG.md [hunk 3]
```diff
@@ -19,12 +19,16 @@ compatible with the updates.
 
 ### Changed
 
-- Refactored BSplines interpolation independetly of the backbone network and available
+- Uniformed local-net, global-net, u-net under a single u-net structure.
```
Comment (YipengHu) [inline CHANGELOG.md:22]: not clear why the unifying is useful or advantageous

File: deepreg/model/backbone/global_net.py [hunk 4]
```diff
@@ -1,17 +1,66 @@
...
+    def get_config(self):
+        """Return the config dictionary for recreating this class."""
+        config = super().get_config()
+        config.update(image_size=self.reference_grid.shape[:3])
+        return config
+
+
 @REGISTRY.register_backbone(name="global")
-class GlobalNet(Backbone):
+class GlobalNet(LocalNet):
```
Comment (YipengHu) [inline deepreg/model/backbone/global_net.py:63]: this inherent is not intuitive and can be confusing - why is this necessary?

**Review by YipengHu — CHANGES_REQUESTED** [review]

File: config/test/affine.yaml (hunk 1 above)
Comment (mathpluscode) [reply config/test/affine.yaml:6]: I've updated the back-compatibility parser, and also the model will calculate depth based on `extract_levels` in case depth is not provided.

So now we are offering more options, without impacting previous behaviors.

Comment (mathpluscode) [reply config/test/affine.yaml:6]: this is the same behaviour as before https://github.com/DeepRegNet/DeepReg/blob/main/deepreg/model/backbone/global_net.py#L64

you encode the image until a certain level, then you take the bottom to produce affine.

If we want more functionalities, it can be addressed in another issue/PR.

File: deepreg/dataset/preprocess.py (hunk 2 above)
Comment (mathpluscode) [reply deepreg/dataset/preprocess.py:348]: emmm I was just moving docs, but this equation is consistent with L339 above? it depends on row or column vectors

File: deepreg/model/backbone/global_net.py (hunk 4 above)
Comment (mathpluscode) [reply deepreg/model/backbone/global_net.py:63]: To simplify the code, technically we can inherit from U-Net instead. But local-net is a u-net anyway

File: CHANGELOG.md (hunk 3 above)
Comment (mathpluscode) [reply CHANGELOG.md:22]: It simplifies the code by a lot, and user can define their own U-Net by changing one component, like the up-sampling / down-sampling, or the conv block via overwriting one single function.

File: deepreg/model/backbone/global_net.py (hunk 4 above)
Comment (mathpluscode) [reply deepreg/model/backbone/global_net.py]: changed to subclassing unet now and added a few comments in docstring to explain

File: config/test/affine.yaml (hunk 1 above)
Comment (YipengHu) [reply config/test/affine.yaml:6]: for local-net, the original behaviours has freedom to use any combinations of these levels, e.g. [0], [0,1,2], [0,2,4], [1,2,3,4], [4]........
for global-net, how is depth or extract_levels relevant?

**Review by YipengHu — CHANGES_REQUESTED** [review]
clarify the extrac_levels argument:
- make sure it has the same flexibility as before;
- make sure it is not required for global-net;
- fine if it is a shortcut for some configurations of different extract_levels, but clarify it
Otherwise happy with the PR.

File: config/test/affine.yaml (hunk 1 above)
Comment (mathpluscode) [reply config/test/affine.yaml:6]: local net still can do this, global net has no need of extraction levels,

The point here is that this PR is not changing the existing behaviours, if you want extract_levels to impact the behavior, it should be in a new issue.

In main, only the max of extract_levels is useful.

**Comment by mathpluscode** [comment]
> clarify the extrac_levels argument:
> 
> * make sure it has the same flexibility as before;
> * make sure it is not required for global-net;
> * fine if it is a shortcut for some configurations of different extract_levels, but clarify it
>   Otherwise happy with the PR.

- extract levels has the same behaviour as before, I don't get it why you think it may not
- extract levels is not mandatory for global-net, you can use depth OR extract levels, the reason to do so if for back compatibility.

What concerns me is that, why you find these not clear? Do we miss some explanation in any class?

File: deepreg/model/backbone/u_net.py [hunk 5]
```diff
@@ -60,61 +74,439 @@ def __init__(
...
+            image_size=image_size,
+            extract_levels=extract_levels,
+            out_channels=out_channels,
+            out_kernel_initializer=out_kernel_initializer,
+            out_activation=out_activation,
+        )
+
+    def call(self, inputs: tf.Tensor, training=None, mask=None) -> tf.Tensor:
+        """
+        Build LocalNet graph based on built layers.
```
Comment (kate-sann5100) [inline deepreg/model/backbone/u_net.py:467]: Build LocalNet graph -> Build UNet graph

**Review by kate-sann5100 — CHANGES_REQUESTED** [review]

**Review by YipengHu — APPROVED** [review]

**Review by kate-sann5100 — APPROVED** [review]

**Comment by zacbaum** [comment]
I know this has been closed now... But having recently pulled from main into my own branch I notice we now call model.summary(). Is this necessary? It's quite a large amount of logging...

**Comment by YipengHu** [comment]
> I know this has been closed now... But having recently pulled from main into my own branch I notice we now call model.summary(). Is this necessary? It's quite a large amount of logging...

I'm for a new ticket to remove if no obvious reason to keep it.

**Comment by mathpluscode** [comment]
> > I know this has been closed now... But having recently pulled from main into my own branch I notice we now call model.summary(). Is this necessary? It's quite a large amount of logging...
> 
> 
> 
> I'm for a new ticket to remove if no obvious reason to keep it.

I guess adding a verbosity option would be nice? So you see these messages only when you are in debug mode.

**Comment by zacbaum** [comment]
> > > I know this has been closed now... But having recently pulled from main into my own branch I notice we now call model.summary(). Is this necessary? It's quite a large amount of logging...
> > 
> > 
> > I'm for a new ticket to remove if no obvious reason to keep it.
> 
> I guess adding a verbosity option would be nice? So you see these messages only when you are in debug mode.

I think that's a perfect solution - this would be good for changing the verbosity of model.fit as well, since we use the default there and that creates massive text files in the process!

**Comment by mathpluscode** [comment]
> > > > I know this has been closed now... But having recently pulled from main into my own branch I notice we now call model.summary(). Is this necessary? It's quite a large amount of logging...
> > > 
> > > 
> > > I'm for a new ticket to remove if no obvious reason to keep it.
> > 
> > 
> > I guess adding a verbosity option would be nice? So you see these messages only when you are in debug mode.
> 
> I think that's a perfect solution - this would be good for changing the verbosity of model.fit as well, since we use the default there and that creates massive logging text files in the process if you are piping that output to a file!

it will be resolved in #614
